File: TypeScript/app/gilded-rose.ts
const MAX_QUALITY = 50;

export class Item {
    name: string;
    sellIn: number;
    quality: number;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

export class GildedRose {
    items: Array<Item>;

    constructor(items = [] as Array<Item>) {
        this.items = items;
    }

    updateQuality() {
        for (let i = 0; i < this.items.length; i++) {
            if (this.items[i].name == 'Sulfuras, Hand of Ragnaros') {
                continue;
            }
            if (this.items[i].name != 'Aged Brie' && this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
                this.reduceQuality(this.items[i]);
            } else {
                this.increaseQualityIfQualityUnderCap(this.items[i]);

                if (this.items[i].name == 'Backstage passes to a TAFKAL80ETC concert') {
                    if (this.items[i].sellIn < 11) {
                        this.increaseQualityIfQualityUnderCap(this.items[i]);
                    }
                    if (this.items[i].sellIn < 6) {
                        this.increaseQualityIfQualityUnderCap(this.items[i]);
                    }
                }

            }
            this.items[i].sellIn = this.items[i].sellIn - 1;
            if (this.items[i].sellIn < 0) {
                if (this.items[i].name != 'Aged Brie') {
                    if (this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
                        this.reduceQuality(this.items[i]);
                    } else {
                        this.items[i].quality = 0;
                    }
                } else {
                    this.increaseQualityIfQualityUnderCap(this.items[i]);
                }
            }
        }

        return this.items;
    }

    reduceQuality(item) {
        if (item.quality > 0) {
            item.quality = item.quality -1;
        }
    }

    increaseQualityIfQualityUnderCap(item) {
        if (item.quality < MAX_QUALITY) {
            item.quality = item.quality + 1;
        }
    }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:

            if item.name == "Sulfuras, Hand of Ragnaros":
                continue

            item.sell_in = item.sell_in - 1

            if item.name != "Aged Brie" and item.name != "Backstage passes to a TAFKAL80ETC concert":
                self.degrade(item)

                if item.sell_in < 0:
                    self.degrade(item)

            # Item is Aged Brie or backstage passes
            if item.name == "Aged Brie" or item.name == "Backstage passes to a TAFKAL80ETC concert":
                if item.quality < 50:  # => 49
                    item.quality = item.quality + 1  # => 50
                    if item.name == "Backstage passes to a TAFKAL80ETC concert":
                        if item.sell_in < 10:
                            if item.quality < 50:
                                item.quality = item.quality + 1
                        if item.sell_in < 5:
                            if item.quality < 50:
                                item.quality = item.quality + 1

            if item.sell_in < 0:
                if item.name == "Backstage passes to a TAFKAL80ETC concert":
                    item.quality = 0
                if item.name == "Aged Brie":
                    if item.quality < 50:
                        item.quality = item.quality + 1

    def degrade(self, item):
        if item.quality > 0:
            item.quality = item.quality - 1


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
